### PR TITLE
support junit in test-cmd

### LINKED ISSUE #45196: cmd/integration tests should produce junit results
/assign @pwittrock @mengqiy 
@kubernetes/sig-cli-misc 

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-test-go/5306?log#log
artifacts link shows that the `make test-cmd` command does not produce junit results: http://gcsweb.k8s.io/gcs/kubernetes-jenkins/logs/ci-kubernetes-test-go/5306/artifacts/

This prevents us from displaying on testgrid what is happening: https://k8s-testgrid.appspot.com/google-unit#test-go

### PR DESCRIPTION
**Pull request by mengqiy**
Output junit result for cmd tests.

Fixes #45196

There will be a followup PR to refactor the cmd test into functions as test cases there is not a lot changes in `hack/make-rules/test-cmd-util.sh`

```release-note
NONE
```

### CODE REVIEW

File: hack/make-rules/test-cmd-util.sh [hunk 1]
```diff
@@ -76,6 +76,43 @@ daemonsets="daemonsets"
...
+# record_command run_kubectl_tests
+#
+# WARNING: Variable changes in the command will NOT be effective after record_command returns.
+#          This is because the command runs in subshell.
+function record_command() {
+    set +o nounset
+    set +o errexit
+
+    local name="$1"
+    local output="/tmp/junit-results"
```
Comment (apelisse) [inline hack/make-rules/test-cmd-util.sh]: You can use `${KUBE_JUNIT_REPORT_DIR:-/tmp/junit-results}` to specify a default value.

File: hack/make-rules/test-cmd-util.sh [hunk 2]
```diff
@@ -76,6 +76,43 @@ daemonsets="daemonsets"
...
+# record_command runs the command and records its output/error messages in junit format
+# it expects the first to be the name of the command
+# Example:
+# record_command run_kubectl_tests
+#
+# WARNING: Variable changes in the command will NOT be effective after record_command returns.
+#          This is because the command runs in subshell.
+function record_command() {
+    set +o nounset
+    set +o errexit
```
Comment (apelisse) [inline hack/make-rules/test-cmd-util.sh]: Why do you need to unset these two options?

I see that you disable it here, and then re-enable inside the test, so I assume it's either in this function or in ` But can you be more explicit?

Comment (mengqiy) [inline hack/make-rules/test-cmd-util.sh]: I need to disable these 2 options to make `juLog` run correctly. And I want these 2 options to be on when running the test cases as what we did before.

Comment (apelisse) [inline hack/make-rules/test-cmd-util.sh]: Can we fix juLog to work with `nounset` and `errexit` instead?

File: hack/make-rules/test-cmd-util.sh [hunk 3]
```diff
@@ -76,6 +76,43 @@ daemonsets="daemonsets"
...
 
 
+# include shell2junit library
+sh2ju="${KUBE_ROOT}/third_party/forked/shell2junit/sh2ju.sh"
+if [[ -f "${sh2ju}" ]]; then
+  source "${sh2ju}"
+else
+  echo "fetching sh2ju.sh from GitHub"
+  source <(curl -fsS --retry 3 'https://raw.githubusercontent.com/kubernetes/kubernetes/master/third_party/forked/shell2junit/sh2ju.sh')
+fi
```
Comment (apelisse) [inline hack/make-rules/test-cmd-util.sh]: This whole `if` thing is very surprising. Why would the file not be here and we would have to get a non-pinned version of it otherwise? What scenario do you have in mind?

File: hack/make-rules/test-cmd-util.sh (hunk 2 above)
Comment (mengqiy) [inline hack/make-rules/test-cmd-util.sh]: I'm trying to do something similar to https://github.com/kubernetes/test-infra/blob/b72c1687c3b71a44bf9953447670fb566ed53f5e/jenkins/e2e-image/e2e-runner.sh#L58-L82.
I will check if there is a fix in [upstream](https://github.com/manolo/shell2junit). If there is, we can bump it?

File: hack/make-rules/test-cmd-util.sh (hunk 3 above)
Comment (mengqiy) [inline hack/make-rules/test-cmd-util.sh]: >What scenario do you have in mind?

I don't know. We were doing this before. Maybe @fejta knows why since his name is in the comment.

File: hack/make-rules/test-cmd-util.sh (hunk 2 above)
Comment (apelisse) [inline hack/make-rules/test-cmd-util.sh]: Yeah, let's forget about that, thanks

File: hack/make-rules/test-cmd-util.sh [hunk 4]
```diff
@@ -3069,42 +3265,45 @@ run_multi_resources_tests() {
...
   kube::log::status "Checking kubectl version"
   kubectl version
 
-  ns_num=0
   create_and_use_new_namespace() {
-    ns_num=$(($ns_num+1))
-    kube::log::status "Creating namespace namespace${ns_num}"
-    kubectl create namespace "namespace${ns_num}"
-    kubectl config set-context "${CONTEXT}" --namespace="namespace${ns_num}"
+    namespace_number=$(date +%s%N)
```
Comment (pwittrock) [inline hack/make-rules/test-cmd-util.sh]: Add a comment why we do this instead of the increment

File: hack/make-rules/test-cmd-util.sh (hunk 3 above)
Comment (pwittrock) [inline hack/make-rules/test-cmd-util.sh]: Lets throw an error for now, and see if it passes the merge queue

File: hack/make-rules/test-cmd-util.sh (hunk 4 above)
Comment (mengqiy) [inline hack/make-rules/test-cmd-util.sh]: Done

**Comment by mengqiy** [comment]
@apelisse @pwittrock PTAL

**Comment by pwittrock** [comment]
@dchen1107 Adding this to the milestone to get finer grain results on the kubectl cmd tests

**Comment by mengqiy** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this
@k8s-bot pull-unit test this

**Comment by mengqiy** [comment]
@k8s-bot test this

File: hack/make-rules/test-cmd-util.sh [hunk 5]
```diff
@@ -237,9 +271,14 @@ setup() {
 # Kubectl version (--short, --client, --output) #
 ########################################################
 run_kubectl_version_tests() {
+  set -o nounset
```
Comment (ixdy) [inline hack/make-rules/test-cmd-util.sh]: is it necessary to do this in every one of these functions?

File: third_party/forked/shell2junit/sh2ju.sh [hunk 6]
```diff
@@ -137,11 +137,6 @@ function juLog() {
   content="$content
     <testcase assertions=\"1\" name=\"$name\" time=\"$time\" classname=\"$class\">
     $failure
-    <system-out>
```
Comment (ixdy) [inline third_party/forked/shell2junit/sh2ju.sh]: why was this removed?

Comment (pwittrock) [inline third_party/forked/shell2junit/sh2ju.sh]: the output from some of the tests had non-printable characters that screwed up the junit results :(

File: hack/make-rules/test-cmd-util.sh (hunk 5 above)
Comment (pwittrock) [inline hack/make-rules/test-cmd-util.sh]: we need to not fail the test if the sh2ju.sh script exits non-0 because that is expected for failed tests.  do you know of a better way of doing this?

File: hack/make-rules/test-cmd-util.sh [hunk 7]
```diff
@@ -77,6 +77,40 @@ daemonsets="daemonsets"
...
+#          This is because the command runs in subshell.
+function record_command() {
+    set +o nounset
+    set +o errexit
+
+    local name="$1"
+    local output="${KUBE_JUNIT_REPORT_DIR:-/tmp/junit-results}"
+    echo "Recording: ${name}"
+    echo "Running command: $@"
+    juLog -output="${output}" -class="test-cmd" -name="${name}" "$@"
```
Comment (ixdy) [inline hack/make-rules/test-cmd-util.sh]: the usual way to handle something returning nonzero with `errexit` is to do something like
```bash
mycmd && rc=$? || rc=$?
if [[ "${rc}" != 0 ]]; then
...
```

File: third_party/forked/shell2junit/sh2ju.sh (hunk 6 above)
Comment (mengqiy) [inline third_party/forked/shell2junit/sh2ju.sh]: No one else is using this now, so this change will not affect anyone else.
And stdout can be found in build log.

Comment (ixdy) [inline third_party/forked/shell2junit/sh2ju.sh]: ah, sounds like https://github.com/kubernetes/test-infra/pull/1839...

File: hack/make-rules/test-cmd-util.sh (hunk 7 above)
Comment (mengqiy) [inline hack/make-rules/test-cmd-util.sh]: It will look like 
```
    juLog -output="${output}" -class="test-cmd" -name="${name}" "$@" && rc=$? || rc=$?
    if [[ "${rc}" != 0 ]]; then
      echo "Error when running ${name}"
      foundError="True"
    fi
```
The first line will become really long. But I will update it as you suggested.

Comment (ixdy) [inline hack/make-rules/test-cmd-util.sh]: yeah, that's pretty much the way shell works. you can always break into several lines with \.

if it can avoid us having the `set -o` / `set +o` stuff everywhere it's an improvement, though.

Comment (ixdy) [inline hack/make-rules/test-cmd-util.sh]: (you're also welcome to do this in a followup PR or something. having junit for test-cmd is a huge improvement, and I don't mean to block that progress.)

Comment (mengqiy) [inline hack/make-rules/test-cmd-util.sh]: I will address the remaining comments in a followup PR. I want to get this PR merged ASAP. Otherwise, I can imagine I will fall into endless rebase when the queue open again.
